README.md
README - Apple CUPS v2.3.6 - 2022-05-25
=======================================

> Note: Apple CUPS is the version of CUPS that is shipped with macOS and iOS.
> For the current version of CUPS that is used on other operating systems, see
> <https://openprinting.github.io/cups> for details.


INTRODUCTION
------------

CUPS is a standards-based, open source printing system developed by Apple Inc.
for macOS® and other UNIX®-like operating systems.  CUPS uses the Internet
Printing Protocol ("IPP") and provides System V and Berkeley command-line
interfaces, a web interface, and a C API to manage printers and print jobs.  It
supports printing to both local (parallel, serial, USB) and networked printers,
and printers can be shared from one computer to another, even over the Internet!

Internally, CUPS uses PostScript Printer Description ("PPD") files to describe
printer capabilities and features and a wide variety of generic and device-
specific programs to convert and print many types of files.  Sample drivers are
included with CUPS to support many Dymo, EPSON, HP, Intellitech, OKIDATA, and
Zebra printers.  Many more drivers are available online and (in some cases) on
the driver CD-ROM that came with your printer.

CUPS is licensed under the Apache License Version 2.0.  See the file
"LICENSE" for more information.


READING THE DOCUMENTATION
-------------------------

Initial documentation to get you started is provided in the root directory of
the CUPS sources:

- `CHANGES.md`: A list of changes in the current major release of CUPS.
- `CONTRIBUTING.md`: Guidelines for contributing to the CUPS project.
- `CREDITS.md`: A list of past contributors to the CUPS project.
- `DEVELOPING.md`: Guidelines for developing code for the CUPS project.
- `INSTALL.md`: Instructions for building and installing CUPS.
- `LICENSE`: The CUPS license agreement (Apache 2.0).
- `NOTICE`: Copyright notices and exceptions to the CUPS license agreement.
- `README.md`: This file.

Once you have installed the software you can access the documentation (and a
bunch of other stuff) online at <http://localhost:631/> and using the `man`
command, for example `man cups`.

If you're having trouble getting that far, the documentation is located under
the `doc/help` and `man` directories.

Please read the documentation before asking questions.


GETTING SUPPORT AND OTHER RESOURCES
-----------------------------------

If you have problems, *read the documentation first!*  We also provide two
mailing lists which are available at <https://lists.cups.org/mailman/listinfo>.

See the CUPS web site at <https://www.cups.org/> for other resources.


SETTING UP PRINTER QUEUES USING YOUR WEB BROWSER
------------------------------------------------

CUPS includes a web-based administration tool that allows you to manage
printers, classes, and jobs on your server.  Open <http://localhost:631/admin/>
in your browser to access the printer administration tools:

*Do not* use the hostname for your machine - it will not work with the default
CUPS configuration.  To enable administration access on other addresses, check
the `Allow Remote Administration` box and click on the `Change Settings` button.

You will be asked for the administration password (root or any other user in the
"sys", "system", "root", "admin", or "lpadmin" group on your system) when
performing any administrative function.


SETTING UP PRINTER QUEUES FROM THE COMMAND-LINE
-----------------------------------------------

CUPS currently uses PPD (PostScript Printer Description) files that describe
printer capabilities and driver programs needed for each printer.  The
`everywhere` PPD is used for nearly all modern networks printers sold since
about 2009.  For example, the following command creates a print queue for a
printer at address "11.22.33.44":

    lpadmin -p printername -E -v ipp://11.22.33.44/ipp/print -m everywhere

CUPS also includes several sample PPD files you can use for "legacy" printers:

   Driver                         | PPD Name
   -----------------------------  | ------------------------------
   Dymo Label Printers            | drv:///sample.drv/dymo.ppd
   Intellitech Intellibar         | drv:///sample.drv/intelbar.ppd
   EPSON 9-pin Series             | drv:///sample.drv/epson9.ppd
   EPSON 24-pin Series            | drv:///sample.drv/epson24.ppd
   Generic PCL Laser Printer      | drv:///sample.drv/generpcl.ppd
   Generic PostScript Printer     | drv:///sample.drv/generic.ppd
   HP DeskJet Series              | drv:///sample.drv/deskjet.ppd
   HP LaserJet Series             | drv:///sample.drv/laserjet.ppd
   OKIDATA 9-Pin Series           | drv:///sample.drv/okidata9.ppd
   OKIDATA 24-Pin Series          | drv:///sample.drv/okidat24.ppd
   Zebra CPCL Label Printer       | drv:///sample.drv/zebracpl.ppd
   Zebra EPL1 Label Printer       | drv:///sample.drv/zebraep1.ppd
   Zebra EPL2 Label Printer       | drv:///sample.drv/zebraep2.ppd
   Zebra ZPL Label Printer        | drv:///sample.drv/zebra.ppd

You can run the `lpinfo -m` command to list all of the available drivers:

    lpinfo -m

Run the `lpinfo -v` command to list the available printers:

    lpinfo -v

Then use the correct URI to add the printer using the `lpadmin` command:

    lpadmin -p printername -E -v device-uri -m ppd-name

Current network printers typically use `ipp` or `ipps` URIS:

    lpadmin -p printername -E -v ipp://11.22.33.44/ipp/print -m everywhere
    lpadmin -p printername -E -v ipps://11.22.33.44/ipp/print -m everywhere

Older network printers typically use `socket` or `lpd` URIs:

    lpadmin -p printername -E -v socket://11.22.33.44 -m ppd-name
    lpadmin -p printername -E -v lpd://11.22.33.44/ -m ppd-name

The sample drivers provide basic printing capabilities, but generally do not
exercise the full potential of the printers or CUPS.  Other drivers provide
greater printing capabilities.


PRINTING FILES
--------------

CUPS provides both the System V `lp` and Berkeley `lpr` commands for printing:

    lp filename
    lpr filename

Both the `lp` and `lpr` commands support printing options for the driver:

    lp -o media=A4 -o resolution=600dpi filename
    lpr -o media=A4 -o resolution=600dpi filename

CUPS recognizes many types of images files as well as PDF, PostScript, and text
files, so you can print those files directly rather than through an application.

If you have an application that generates output specifically for your printer
then you need to use the `-oraw` or `-l` options:

    lp -o raw filename
    lpr -l filename

This will prevent the filters from misinterpreting your print file.


LEGAL STUFF
-----------

Copyright © 2007-2021 by Apple Inc.
Copyright © 1997-2007 by Easy Software Products.

CUPS is provided under the terms of the Apache License, Version 2.0 with
exceptions for GPL2/LGPL2 software.  A copy of this license can be found in the
file `LICENSE`.  Additional legal information is provided in the file `NOTICE`.

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied.  See the License for the
specific language governing permissions and limitations under the License.


CHANGES.md
Change History
==============

Changes in CUPS v2.3.6
----------------------
- CVE-2022-26691: An incorrect comparison in local admin authentication.


Changes in CUPS v2.3.5
----------------------

- The automated test suite can now be activated using `make test` for
  consistency with other projects and CI environments - the old `make check`
  continues to work as well, and the previous test server behavior can be
  accessed by running `make testserver`.
- ippeveprinter now supports multiple icons and strings files.
- ippeveprinter now uses the system's FQDN with Avahi.
- ippeveprinter now supports Get-Printer-Attributes on "/".
- ippeveprinter now uses a deterministic "printer-uuid" value.
- ippeveprinter now uses system sounds on macOS for Identify-Printer.
- Updated ippfind to look for files in "~/Desktop" on Windows.
- Updated ippfind to honor `SKIP-XXX` directives with `PAUSE`.
- Updated IPP Everywhere support to work around printers that only advertise
  color raster support but really also support grayscale (OpenPrinting #1)
- ipptool now supports DNS-SD URIs like `ipps://My%20Printer._ipps._tcp.local`
  (OpenPrinting #5)
- ipptool now supports monitoring the printer state while submitting a job
  with the `MONITOR-PRINTER-STATE` directive (OpenPrinting #153)
- ipptool now supports testing for unique values with the `WITH-DISTINCT-VALUES`
  predicate (OpenPrinting #153)
- ipptool now supports retrying requests on a `server-error-busy` status code
  (OpenPrinting #153)
- ipptool now supports `value-tag(MAX)` and `value-tag(MIN:MAX)` for the
  `OF-TYPE` predicate (OpenPrinting #153)
- The scheduler now allows root backends to have world read permissions but not
  world execute permissions (OpenPrinting #21)
- Failures to bind IPv6 listener sockets no longer cause errors if IPv6 is
  disabled on the host (OpenPrinting #25)
- The SNMP backend now supports the HP and Ricoh vendor MIBs (OpenPrinting #28)
- The scheduler no longer includes a timestamp in files it writes (OpenPrinting #29)
- IPP Everywhere PPDs could have an "unknown" default InputSlot (OpenPrinting #44)
- The `httpAddrListen` function now uses a listen backlog of 128.
- The PPD functions now treat boolean values as case-insensitive (OpenPrinting #106)
- Temporary queue names no longer end with an underscore (OpenPrinting #110)
- Added USB quirks (Issue #5789, #5766, #5823, #5831, #5838, #5843, #5867)
- Fixed IPP Everywhere v1.1 conformance issues in ippeveprinter.
- Fixed DNS-SD name collision support in ippeveprinter.
- Fixed compiler and code analyzer warnings.
- Fixed TLS support on Windows.
- Fixed ippfind sub-type searches with Avahi.
- Fixed the default hostname used by ippeveprinter on macOS.
- Fixed resolution of local IPP-USB printers with Avahi.
- Fixed coverity issues (OpenPrinting #2)
- Fixed `httpAddrConnect` issues (OpenPrinting #3)
- Fixed web interface device URI issue (OpenPrinting #4)
- Fixed lp/lpr "printer/class not found" error reporting (OpenPrinting #6)
- Fixed a memory leak in the scheduler (OpenPrinting #12)
- Fixed a potential integer overflow in the PPD hashing code (OpenPrinting #13)
- Fixed output-bin and print-quality handling issues (OpenPrinting #18)
- Fixed PPD options getting mapped to odd IPP values like "tray---4" (OpenPrinting #23)
- Fixed remote access to the cupsd.conf and log files (OpenPrinting #24)
- Fixed a logging regression caused by a previous change for Issue #5604
  (OpenPrinting #25)
- Fixed the "uri-security-supported" value from the scheduler (OpenPrinting #42)
- Fixed IPP backend crash bug with "printer-alert" values (OpenPrinting #43)
- Fixed default options that incorrectly use the "custom" prefix (OpenPrinting #48)
- Fixed a memory leak when resolving DNS-SD URIs (OpenPrinting #49)
- Fixed cupsManualCopies values in IPP Everywhere PPDs (Issue #5807)
- Fixed duplicate ColorModel entries for AirPrint printers (Issue 59)
- Fixed crash bug in `ppdOpen` (OpenPrinting #64, OpenPrinting #78)
- Fixed regression in `snprintf` emulation function (OpenPrinting #67)
- Fixed reporting of printer instances when enumerating and when no options are
  set for the main instance (OpenPrinting #71)
- Fixed segfault in help.cgi when searching in man pages (OpenPrinting #81)
- Fixed a bug in ipptool that caused the reuse of request IDs when repeating a
  test (OpenPrinting #153)
- Root certificates were incorrectly stored in "~/.cups/ssl".
- Fixed a PPD memory leak caused by emulator definitions (OpenPrinting #124)
- Fixed a `DISPLAY` bug in `ipptool` (OpenPrinting #139)
- `httpReconnect2` did not reset the socket file descriptor when the TLS
  negotiation failed (Issue #5907)
- `httpUpdate` did not reset the socket file descriptor when the TLS
  negotiation failed (Apple #5915)
- The `ippeveprinter` tool now automatically uses an available port.
- The IPP backend now retries Validate-Job requests (OpenPrinting #132)
- Removed support for the (long deprecated and unused) `KeepAliveTimeout`
  directive in `cupsd.conf` (Issue #5733)
- Fixed `@IF(name)` handling in `cupsd.conf` (Issue #5918)
- The scheduler now supports the "everywhere" model directly (Issue #5919)
- Fixed documentation and added examples for CUPS' limited CGI support
  (Issue #5940)
- Fixed the `lpc` command prompt (Issue #5946)
- Fixed `job-pages-per-set` value for duplex print jobs.


Changes in CUPS v2.3.4
----------------------

- CVE-2020-10001: Fixed a buffer (read) overflow in the `ippReadIO` function.


Changes in CUPS v2.3.3
----------------------

- CVE-2020-3898: The `ppdOpen` function did not handle invalid UI
  constraint.  `ppdcSource::get_resolution` function did not handle
  invalid resolution strings.
- CVE-2019-8842: The `ippReadIO` function may under-read an extension
  field.
- Fixed WARNING_OPTIONS support for GCC 9.x


Changes in CUPS v2.3.2
----------------------

- Localization updates.


Changes in CUPS v2.3.1
----------------------

- Documentation updates (Issue #5661, #5674, #5682)

cups/md5.c
/*
 * Private MD5 implementation for CUPS.
 *
 * Copyright 2007-2017 by Apple Inc.
 * Copyright 2005 by Easy Software Products
 * Copyright (C) 1999 Aladdin Enterprises.  All rights reserved.
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 *
 * L. Peter Deutsch
 * ghost@aladdin.com
 */
/*
  Independent implementation of MD5 (RFC 1321).

  This code implements the MD5 Algorithm defined in RFC 1321.
  It is derived directly from the text of the RFC and not from the
  reference implementation.

  The original and principal author of md5.c is L. Peter Deutsch
  <ghost@aladdin.com>.  Other authors are noted in the change history
  that follows (in reverse chronological order):

  1999-11-04 lpd Edited comments slightly for automatic TOC extraction.
  1999-10-18 lpd Fixed typo in header comment (ansi2knr rather than md5).
  1999-05-03 lpd Original version.
 */

#include "md5-internal.h"
#include "string-private.h"

#if !defined(__APPLE__)
#  define T1 0xd76aa478
#  define T2 0xe8c7b756
#  define T3 0x242070db
#  define T4 0xc1bdceee
#  define T5 0xf57c0faf
#  define T6 0x4787c62a
#  define T7 0xa8304613
#  define T8 0xfd469501
#  define T9 0x698098d8
#  define T10 0x8b44f7af
#  define T11 0xffff5bb1
#  define T12 0x895cd7be
#  define T13 0x6b901122
#  define T14 0xfd987193
#  define T15 0xa679438e
#  define T16 0x49b40821
#  define T17 0xf61e2562
#  define T18 0xc040b340
#  define T19 0x265e5a51
#  define T20 0xe9b6c7aa
#  define T21 0xd62f105d
#  define T22 0x02441453
#  define T23 0xd8a1e681
#  define T24 0xe7d3fbc8
#  define T25 0x21e1cde6
#  define T26 0xc33707d6
#  define T27 0xf4d50d87
#  define T28 0x455a14ed
#  define T29 0xa9e3e905
#  define T30 0xfcefa3f8
#  define T31 0x676f02d9
#  define T32 0x8d2a4c8a
#  define T33 0xfffa3942
#  define T34 0x8771f681
#  define T35 0x6d9d6122
#  define T36 0xfde5380c
#  define T37 0xa4beea44
#  define T38 0x4bdecfa9
#  define T39 0xf6bb4b60
#  define T40 0xbebfbc70
#  define T41 0x289b7ec6
#  define T42 0xeaa127fa
#  define T43 0xd4ef3085
#  define T44 0x04881d05
#  define T45 0xd9d4d039
#  define T46 0xe6db99e5
#  define T47 0x1fa27cf8
#  define T48 0xc4ac5665
#  define T49 0xf4292244
#  define T50 0x432aff97
#  define T51 0xab9423a7
#  define T52 0xfc93a039
#  define T53 0x655b59c3
#  define T54 0x8f0ccc92
#  define T55 0xffeff47d
#  define T56 0x85845dd1
#  define T57 0x6fa87e4f
#  define T58 0xfe2ce6e0
#  define T59 0xa3014314
#  define T60 0x4e0811a1
#  define T61 0xf7537e82
#  define T62 0xbd3af235
#  define T63 0x2ad7d2bb
#  define T64 0xeb86d391

static void
_cups_md5_process(_cups_md5_state_t *pms, const unsigned char *data /*[64]*/)
{
    unsigned int
	a = pms->abcd[0], b = pms->abcd[1],
	c = pms->abcd[2], d = pms->abcd[3];
    unsigned int t;

#  ifndef ARCH_IS_BIG_ENDIAN
#    define ARCH_IS_BIG_ENDIAN 1	/* slower, default implementation */
#  endif
#  if ARCH_IS_BIG_ENDIAN

    /*
     * On big-endian machines, we must arrange the bytes in the right
     * order.  (This also works on machines of unknown byte order.)
     */
    unsigned int X[16];
    const unsigned char *xp = data;
    int i;

    for (i = 0; i < 16; ++i, xp += 4)
	X[i] = (unsigned)xp[0] + ((unsigned)xp[1] << 8) +
	       ((unsigned)xp[2] << 16) + ((unsigned)xp[3] << 24);

#  else  /* !ARCH_IS_BIG_ENDIAN */

    /*
     * On little-endian machines, we can process properly aligned data
     * without copying it.
     */
    unsigned int xbuf[16];
    const unsigned int *X;

    if (!((data - (const unsigned char *)0) & 3)) {
	/* data are properly aligned */
	X = (const unsigned int *)data;
    } else {
	/* not aligned */
	memcpy(xbuf, data, 64);
	X = xbuf;
    }
#  endif

#  define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32 - (n))))

    /* Round 1. */
    /* Let [abcd k s i] denote the operation
       a = b + ((a + F(b,c,d) + X[k] + T[i]) <<< s). */
#  define F(x, y, z) (((x) & (y)) | (~(x) & (z)))
#  define SET(a, b, c, d, k, s, Ti)\
  t = a + F(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
    /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  7,  T1);
    SET(d, a, b, c,  1, 12,  T2);
    SET(c, d, a, b,  2, 17,  T3);
    SET(b, c, d, a,  3, 22,  T4);
    SET(a, b, c, d,  4,  7,  T5);
    SET(d, a, b, c,  5, 12,  T6);
    SET(c, d, a, b,  6, 17,  T7);
    SET(b, c, d, a,  7, 22,  T8);
    SET(a, b, c, d,  8,  7,  T9);
    SET(d, a, b, c,  9, 12, T10);
    SET(c, d, a, b, 10, 17, T11);
    SET(b, c, d, a, 11, 22, T12);
    SET(a, b, c, d, 12,  7, T13);
    SET(d, a, b, c, 13, 12, T14);
    SET(c, d, a, b, 14, 17, T15);
    SET(b, c, d, a, 15, 22, T16);
#  undef SET

     /* Round 2. */
     /* Let [abcd k s i] denote the operation
          a = b + ((a + G(b,c,d) + X[k] + T[i]) <<< s). */
#  define G(x, y, z) (((x) & (z)) | ((y) & ~(z)))
#  define SET(a, b, c, d, k, s, Ti)\
  t = a + G(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  1,  5, T17);
    SET(d, a, b, c,  6,  9, T18);
    SET(c, d, a, b, 11, 14, T19);
    SET(b, c, d, a,  0, 20, T20);
    SET(a, b, c, d,  5,  5, T21);
    SET(d, a, b, c, 10,  9, T22);
    SET(c, d, a, b, 15, 14, T23);
    SET(b, c, d, a,  4, 20, T24);
    SET(a, b, c, d,  9,  5, T25);
    SET(d, a, b, c, 14,  9, T26);
    SET(c, d, a, b,  3, 14, T27);
    SET(b, c, d, a,  8, 20, T28);
    SET(a, b, c, d, 13,  5, T29);
    SET(d, a, b, c,  2,  9, T30);
    SET(c, d, a, b,  7, 14, T31);
    SET(b, c, d, a, 12, 20, T32);
#  undef SET

     /* Round 3. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + H(b,c,d) + X[k] + T[i]) <<< s). */
#  define H(x, y, z) ((x) ^ (y) ^ (z))
#  define SET(a, b, c, d, k, s, Ti)\
  t = a + H(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  5,  4, T33);
    SET(d, a, b, c,  8, 11, T34);
    SET(c, d, a, b, 11, 16, T35);
    SET(b, c, d, a, 14, 23, T36);
    SET(a, b, c, d,  1,  4, T37);
    SET(d, a, b, c,  4, 11, T38);
    SET(c, d, a, b,  7, 16, T39);
    SET(b, c, d, a, 10, 23, T40);
    SET(a, b, c, d, 13,  4, T41);
    SET(d, a, b, c,  0, 11, T42);
    SET(c, d, a, b,  3, 16, T43);
    SET(b, c, d, a,  6, 23, T44);
    SET(a, b, c, d,  9,  4, T45);
    SET(d, a, b, c, 12, 11, T46);
    SET(c, d, a, b, 15, 16, T47);
    SET(b, c, d, a,  2, 23, T48);
#  undef SET

     /* Round 4. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + I(b,c,d) + X[k] + T[i]) <<< s). */
#  define I(x, y, z) ((y) ^ ((x) | ~(z)))
#  define SET(a, b, c, d, k, s, Ti)\
  t = a + I(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  6, T49);
    SET(d, a, b, c,  7, 10, T50);
    SET(c, d, a, b, 14, 15, T51);
    SET(b, c, d, a,  5, 21, T52);
    SET(a, b, c, d, 12,  6, T53);
    SET(d, a, b, c,  3, 10, T54);
    SET(c, d, a, b, 10, 15, T55);
    SET(b, c, d, a,  1, 21, T56);
    SET(a, b, c, d,  8,  6, T57);
    SET(d, a, b, c, 15, 10, T58);
    SET(c, d, a, b,  6, 15, T59);
    SET(b, c, d, a, 13, 21, T60);
    SET(a, b, c, d,  4,  6, T61);
    SET(d, a, b, c, 11, 10, T62);
    SET(c, d, a, b,  2, 15, T63);
    SET(b, c, d, a,  9, 21, T64);
#  undef SET

     /* Then perform the following additions. (That is increment each
        of the four registers by the value it had before this block
        was started.) */
    pms->abcd[0] += a;
    pms->abcd[1] += b;
    pms->abcd[2] += c;
    pms->abcd[3] += d;
}

void
_cupsMD5Init(_cups_md5_state_t *pms)
{
    pms->count[0] = pms->count[1] = 0;
    pms->abcd[0] = 0x67452301;
    pms->abcd[1] = 0xefcdab89;
    pms->abcd[2] = 0x98badcfe;
    pms->abcd[3] = 0x10325476;
}

void
_cupsMD5Append(_cups_md5_state_t *pms, const unsigned char *data, int nbytes)
{
    const unsigned char *p = data;
    int left = nbytes;
    int offset = (pms->count[0] >> 3) & 63;
    unsigned int nbits = (unsigned int)(nbytes << 3);

    if (nbytes <= 0)
	return;

    /* Update the message length. */
    pms->count[1] += (unsigned)nbytes >> 29;
    pms->count[0] += nbits;
    if (pms->count[0] < nbits)
	pms->count[1]++;

    /* Process an initial partial block. */
    if (offset) {
	int copy = (offset + nbytes > 64 ? 64 - offset : nbytes);

	memcpy(pms->buf + offset, p, (size_t)copy);
	if (offset + copy < 64)
	    return;
	p += copy;
	left -= copy;
	_cups_md5_process(pms, pms->buf);
    }

    /* Process full blocks. */
    for (; left >= 64; p += 64, left -= 64)
	_cups_md5_process(pms, p);

    /* Process a final partial block. */
    if (left)
	memcpy(pms->buf, p, (size_t)left);
}

void
_cupsMD5Finish(_cups_md5_state_t *pms, unsigned char digest[16])
{
    static const unsigned char pad[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };
    unsigned char data[8];
    int i;

    /* Save the length before padding. */
    for (i = 0; i < 8; ++i)
	data[i] = (unsigned char)(pms->count[i >> 2] >> ((i & 3) << 3));
    /* Pad to 56 bytes mod 64. */
    _cupsMD5Append(pms, pad, (int)((55 - (pms->count[0] >> 3)) & 63) + 1);
    /* Append the length. */
    _cupsMD5Append(pms, data, 8);
    for (i = 0; i < 16; ++i)
	digest[i] = (unsigned char)(pms->abcd[i >> 2] >> ((i & 3) << 3));
}
#endif /* !__APPLE__ */
